projects/meta_cl/networks/dendritic_networks.py

```diff
@@ -184,6 +182,116 @@ def conv_block(cls, in_channels, out_channels, kernel_size, stride, padding):
                 in_channels=in_channels, out_channels=out_channels,
                 kernel_size=kernel_size, stride=stride, padding=padding
             ),
+            nn.InstanceNorm2d(num_features=out_channels, affine=True),
+            nn.ReLU(),
+        ]
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """
+        Apply dendrites as a gating mechanism.
+
+        TODO: Decide best method to apply dendrites.
+        """
+        # # Multiple by the sigmoid of the max along each segment.
+        # return y * torch.sigmoid(dendrite_activations.max(dim=2).values)
+
+        inds = dendrite_activations.abs().max(dim=2).indices
+        inds = inds.unsqueeze(dim=2)
+        dendrite_activations = torch.gather(dendrite_activations, dim=2, index=inds)
+        dendrite_activations = dendrite_activations.squeeze(dim=2)
+        dendrite_activations = torch.sigmoid(dendrite_activations)
+        return y * dendrite_activations
+
+    def forward(self, x):
+
+        mod = self.modulation(x)
+        pred = self.prediction(x)
+
+        dendrite_activations = self.segments(mod)
+        out = self.apply_dendrites(pred, dendrite_activations)
+
+        out = self.classifier(out)
+
+        return out
+
+
+class CloserToANMLDendriticNetwork(nn.Module):
+    """
+    An iteration of a dendritic network, based closely on `ANML`_. (The conv + max-pool
+    layers are identical). It's similar to ANMLDendriticNetwork, but does not include a
+    linear layer atop the modulation network. This makes it more similar to ANML's
+    original network. In fact, with `num_segments=1, dendrite_sparsity=0`, they should
```

Comment:
Is it correct to conclude that they are? Besides the empirical results with the full training, did you also look at the networks via other means and came to that same conclusion?

Reply:
I did not. I didn't want to spend too much time on this considering it was just a sanity check. Manual inspection shows that the architectures are identical baring the linear layers swapped for the dendrites layer, so I left it there.